### DISCUSSION TITLE
Format on save for typescript files

### DISCUSSION THREAD

**Discussion by ukcw** [LazyVim/LazyVim#242] [Q&A] [answered]
I've been trying to make nvim format my `.tsx` file on save by installing the prettier formatter but it doesn't work. I'm not sure how to achieve this so any help is appreciated!

<img width="1390" alt="image" src="https://user-images.githubusercontent.com/34738958/218266905-2e0e074e-56ff-480c-8c23-17ee110b2ae9.png">
This is my `:LspInfo` output.

**Comment by folke** [comment]
Looking at your screenshot, it seems you don't have null-ls enabled? Or not configured to use prettier?

Check the null-ls docs to see how to configure it

**Comment by pprotas** [comment, marked as answer]
The configuration for this will depend on your project. Here are some options that I recommend (from best to worst):
1. Use ESLint and Prettier together to format
2. Use the `tsserver` formatter
3. Use `prettier_d` with `null-ls`

Option 1 is the fastest and uses Prettier in combination with ESLint. The second option is also very fast, but it doesn't use Prettier but the TSServer formatter, which is not as good. The third option I only recommend if you REALLY want to use Prettier, but don't have ESLint set up for your project (less effort).

## Option 1
For option 1, you have to disable the TSServer formatter and tell nvim to use eslint instead. There might be a nicer/cleaner way to do this (so take this config with a grain of salt), but this is what worked for me:
```lua
-- in lua/plugins/lsp.lua
return {
  {
    "neovim/nvim-lspconfig",
    opts = {
      setup = {
        tsserver = function(_, opts)
          require("lazyvim.util").on_attach(function(client, buffer)
            if client.name == "tsserver" then
              client.server_capabilities.documentFormattingProvider = false
            end
            if client.name == "eslint" then
              client.server_capabilities.documentFormattingProvider = true
            end
          end)
          require("typescript").setup({ server = opts })
          return true
        end,
      },
    },
  },
  {
    "williamboman/mason.nvim",
    opts = {
      ensure_installed = {
        "eslint-lsp",
      },
    },
  },
},
```

### Neoconf
In your project, use the Neoconf plugin (installed by default) to enable ESLint formatting and tell ESLint where to look:
```json
// In <your project folder>/.neoconf.json
{
  "lspconfig": {
    "eslint": {
      "nodePath": "frontend", // Only add this if your front-end code is in a different folder
      "format": {
        "enable": true
      }
    }
  }
}
```

### ESLint Prettier plugin
Make sure to configure [the ESLint Prettier plugin](https://github.com/prettier/eslint-plugin-prettier)

## Option 2
Option 2 should already work since you are using `tsserver`.

## Option 3
Use `null-ls` to format your files using `prettier-d`. I don't recommend this option since it is very slow, but it is the easiest way to set up Prettier formatting.

```lua
-- in lua/plugins/lsp.lua
return {
  {
    "williamboman/mason.nvim",
    opts = {
      ensure_installed = {
        "prettierd",
      },
    },
  },
  {
    "jose-elias-alvarez/null-ls.nvim",
    opts = function()
      local null_ls = require("null-ls")
      return {
        sources = {
          null_ls.builtins.formatting.prettierd,
          -- All other sources
        },
      }
    end,
  },
}
```

You might have to disable `tsserver` like in option 1, since there might be conflicts in the formatters.

**Comment by ukcw** [comment]
Thanks for the answer @folke and @pprotas! Right now, I've activated the typescript modules within the original `lazy.lua` file and things seem to be working so far. I'll come back to this if I face any new issues. Thanks again!
<img width="962" alt="image" src="https://user-images.githubusercontent.com/34738958/218635983-ae648c06-56d6-4d85-a2cd-bee021f6728f.png">

Seems like activating the typescript modules also enabled `null-ls`?
<img width="1728" alt="image" src="https://user-images.githubusercontent.com/34738958/218636562-b409e335-a14a-4695-aa24-3e2dec292647.png">

**Comment by ramedov** [comment]
For future googlers, if you want to simply disable `tsserver` formatting, here's modified pprotas's solution that worked for me:
```lua
return {
    {
        "neovim/nvim-lspconfig",
         opts = {
            setup = {
                tsserver = function()
                    require("lazyvim.util").lsp.on_attach(function(client)
                        if client.name == "tsserver" then
                            client.server_capabilities.documentFormattingProvider = false
                        end
                    end)
                end,
            },
        },
    },
}
```

**Reply by alewis001** [reply]
Thanks @rismailov, I used the same approach with Volar and it worked for me. What's interesting is that the solution differs to what is described here: https://www.lazyvim.org/plugins/lsp#nvim-lspconfig, in the comments within the `setup` section. I tried following those instructions but got a "module not defined" error. 

So either the docs are wrong or, I was doing something wrong when trying to follow the example.

**Reply by alewis001** [reply]
Actually, I'll ask a new question
